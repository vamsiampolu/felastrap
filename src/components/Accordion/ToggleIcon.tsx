import type { IStyle } from "fela";
import { useFela } from "react-fela";

const iconRule = ({ isOpen = false }) => {
  const style: IStyle = {
    flexShrink: 0,
    width: "1.25rem",
    height: "1.25rem",
    marginLeft: "auto",
    backgroundImage: 'url("/chevron-down.svg")',
    backgroundRepeat: "no-repeat",
    backgroundSize: "1.25rem",
    transition: "transform 0.2s ease-in-out",
    "@media (prefers-reduced-motion: reduce)": {
      transition: "none",
    },
  };

  if (isOpen) {
    style.transform = "rotate(-180deg)";
  }

  return style;
};

export const ToggleIcon = ({ isOpen = false }: { isOpen?: boolean }) => {
  const { css } = useFela({ isOpen });
  const cx = `${css(iconRule)} accordion-toggle-icon`;
  return <div data-testid="accordion-toggle-icon" className={cx} />;
};
