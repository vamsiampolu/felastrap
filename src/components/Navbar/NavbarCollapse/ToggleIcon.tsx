import { useFela } from "react-fela";
import toggleIcon from "/toggle-icon.svg";
import { NavbarVariant } from "../Navbar";
import { NavbarSize, sizeBreakpointMapping } from "../utils";
import { mediaQueryByName } from "../../Flex/mediaQuery";
import { StyleProps } from "../../common";

const base = () => ({
  display: "inline-block",
  width: "1.5em",
  height: "1.5em",
  verticalAlign: "middle",
  backgroundImage: `url(${toggleIcon})`,
  backgroundRepeat: "no-repeat",
  backgroundPosition: "center",
  backgroundSize: "100%",
});

const expandSizeRule = ({ size, variant }: StyleProps<ToggleIconProps>) => {
  const breakpoint = sizeBreakpointMapping[size as NavbarSize];
  if (variant === "expand") {
    return {
      [mediaQueryByName[breakpoint]]: { display: "none " },
    };
  }

  return {};
};

export type ToggleIconProps = {
  className?: string;
  size?: NavbarSize;
  variant?: NavbarVariant;
};

export const ToggleIcon = ({
  className = "",
  size = "medium",
  variant = "expand",
}: ToggleIconProps) => {
  const { css } = useFela({ size, variant });
  const cx = `${css([
    base,
    expandSizeRule,
  ])} navbar-toggle-icon ${className}`.trim();
  return <span data-testid="navbar-toggle-icon" className={cx} />;
};
