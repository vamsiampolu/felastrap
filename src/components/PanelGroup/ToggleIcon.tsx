import { useFela } from "react-fela";
import { StyleProps } from "../common";

const rule = () => ({
  flexShrink: 0,
  width: "1.25rem",
  height: "1.25rem",
  marginLeft: "auto",
  backgroundImage: 'url("/chevron-down.svg")',
  backgroundRepeat: "no-repeat",
  backgroundSize: "1.25rem",
  transition: "transform 0.2s ease-in-out",
  "@media (prefers-reduced-motion: reduce)": {
    transition: "none",
  },
});

const openRule = ({ isOpen }: StyleProps<{ isOpen: boolean }>) =>
  isOpen ? { transform: "rotate(-180deg)" } : {};

export const ToggleIcon = ({ isOpen = false }: { isOpen?: boolean }) => {
  const { css } = useFela({ isOpen });
  const cx = css([rule, openRule]);
  return <div data-testid="panel-group-toggle-icon" className={cx} />;
};
