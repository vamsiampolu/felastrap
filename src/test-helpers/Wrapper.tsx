import type { IRenderer } from "fela";
import { type PropsWithChildren } from "react";
import { RendererProvider, ThemeProvider } from "react-fela";

const Wrapper = ({
  renderer,
  theme = {},
  children,
}: PropsWithChildren<{ renderer: IRenderer; theme?: object }>) => (
  <ThemeProvider theme={theme}>
    <RendererProvider renderer={renderer}>{children}</RendererProvider>;
  </ThemeProvider>
);

export { Wrapper };
