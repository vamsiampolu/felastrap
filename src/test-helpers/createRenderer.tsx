import expandShorthand from "fela-plugin-expand-shorthand";
import sortClassNames from "fela-sort-classnames";
import sortMediaQueryMobileFirst from "fela-sort-media-query-mobile-first";
import prefixer from "fela-plugin-prefixer";
import placeholderPrefixer from "fela-plugin-placeholder-prefixer";
import { createRenderer as createFelaRenderer, type IRenderer } from "fela";
import { lightTheme } from "../reboot/lightTheme";

export const createRenderer = () => {
  const autoMergeLonghands = true;
  const enhancers = [sortClassNames(), sortMediaQueryMobileFirst()];
  const plugins = [
    prefixer(),
    placeholderPrefixer(),
    expandShorthand(autoMergeLonghands),
  ];

  return createFelaRenderer({
    plugins,
    enhancers,
  });
};

export const cssReboot = (renderer: IRenderer) => {
  renderer.renderStatic(lightTheme);
};
