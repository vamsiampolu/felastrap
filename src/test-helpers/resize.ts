export const resizeWidth = (width: number) => {
  window.innerWidth = width;
  dispatchEvent(new Event("resize"));
};

export const resizeHeight = (height: number) => {
  window.innerHeight = height;
  dispatchEvent(new Event("resize"));
};
