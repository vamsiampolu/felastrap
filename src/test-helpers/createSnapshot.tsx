import { IRenderer } from "fela";
import { renderToString } from "fela-tools";
import { ReactElement, ReactNode } from "react";

import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import {
  type UserEvent,
  type UserEventApi,
} from "@testing-library/user-event/dist/types/setup/setup";
import * as prettier from "prettier";
import { renderToStaticMarkup } from "react-dom/server";
import { Wrapper } from "./Wrapper";

async function formatCSS(css: string) {
  return prettier.format(css, {
    parser: "css",
    useTabs: false,
    tabWidth: 2,
  });
}

function formatHTML(str: string) {
  return prettier.format(str, { parser: "html" });
}

function renderToMarkup(component: ReactElement): string {
  return renderToStaticMarkup(component);
}

export const createSnapshot = async ({
  component,
  renderer,
  theme = {},
}: {
  component: ReactNode;
  renderer: IRenderer;
  theme?: object;
}) => {
  // reset renderer to have a clean setup
  renderer.clear();

  const element = (
    <Wrapper theme={theme} renderer={renderer}>
      {component}
    </Wrapper>
  );

  const view = renderToMarkup(element);
  const str = renderToString(renderer);

  const css = await formatCSS(str);
  const html = await formatHTML(view);

  return `${css}\n\n${html}`;
};

export type EventName = keyof UserEventApi | "focus" | "focus-visible";
export type CreateUserEventType = {
  eventName: EventName;
  component: ReactElement;
  renderer: IRenderer;
  testId: string;
  theme?: object;
  donotWrap?: boolean;
};

export type SetupEvent = {
  user: UserEvent;
  eventName: EventName;
  element: HTMLElement;
};
const setupEvent = async ({ user, eventName, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.hover(element);
      break;
    }

    case "focus": {
      element.focus();
      break;
    }

    case "focus-visible": {
      element.contentEditable = "true";
      element.focus();
      element.contentEditable = "false";
      break;
    }
  }
};

const clearEvent = async ({ eventName, user, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.unhover(element);
      break;
    }

    case "focus":
    case "focus-visible": {
      element.blur();
      break;
    }
  }
};

export const createUserEventSnapshot = async ({
  eventName,
  component,
  testId,
  renderer,
  theme = {},
  donotWrap = false,
}: CreateUserEventType) => {
  renderer.clear();

  const user = userEvent.setup();
  let wrappedComponent = component;
  if (!donotWrap) {
    wrappedComponent = (
      <Wrapper theme={theme} renderer={renderer}>
        {component}
      </Wrapper>
    );
  }

  render(wrappedComponent);

  const element = await screen.findByTestId(testId);

  setupEvent({ element, eventName, user });
  const view = await renderToMarkup(wrappedComponent);
  const utils = renderToString(renderer);
  const css = await formatCSS(utils);
  const html = await formatHTML(view);

  clearEvent({ eventName, element, user });

  return `${css}\n\n${html}`;
};
