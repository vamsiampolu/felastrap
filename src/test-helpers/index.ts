import { Wrapper } from "./Wrapper";
import { createRenderer, cssReboot } from "./createRenderer";
import { createSnapshot, createUserEventSnapshot } from "./createSnapshot";
import { resizeHeight, resizeWidth } from "./resize";

export {
  createRenderer,
  cssReboot,
  createSnapshot,
  createUserEventSnapshot,
  resizeWidth,
  resizeHeight,
  Wrapper,
};
