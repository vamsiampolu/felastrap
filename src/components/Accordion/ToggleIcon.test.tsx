import { IRenderer } from "fela";
import { ToggleIcon } from "./ToggleIcon";
import { Wrapper, createRenderer, createSnapshot } from "../../test-helpers";
import { render, screen } from "@testing-library/react";

describe("components/Accordion/ToggleIcon", () => {
  let renderer: IRenderer;

  beforeAll(() => {
    renderer = createRenderer();
  });

  beforeEach(() => {
    renderer.clear();
  });

  afterAll(() => {
    renderer = undefined as unknown as IRenderer;
  });

  it("should render a ToggleIcon", async () => {
    const component = <ToggleIcon />;
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("accordion-toggle-icon")).toBeInTheDocument();
    expect(screen.queryByTestId("accordion-toggle-icon")).toHaveClass(
      "accordion-toggle-icon",
    );

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render an open ToggleIcon", async () => {
    const component = <ToggleIcon isOpen />;
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("accordion-toggle-icon")).toBeInTheDocument();
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });
});
