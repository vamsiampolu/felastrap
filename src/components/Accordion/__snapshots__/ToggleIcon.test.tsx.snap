// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Accordion/ToggleIcon should render a ToggleIcon 1`] = `
".a {
  flex-shrink: 0;
}
.b {
  width: 1.25rem;
}
.c {
  height: 1.25rem;
}
.d {
  margin-left: auto;
}
.e {
  background-image: url("/chevron-down.svg");
}
.f {
  background-repeat: no-repeat;
}
.g {
  background-size: 1.25rem;
}
.l {
  transition-property: transform;
}
.m {
  transition-duration: 0.2s;
}
.n {
  transition-delay: ease-in-out;
}
.o {
  transition-timing-function: ease;
}
@media (prefers-reduced-motion: reduce) {
  .h {
    transition-property: none;
  }
  .i {
    transition-duration: 0s;
  }
  .j {
    transition-delay: 0s;
  }
  .k {
    transition-timing-function: ease;
  }
}


<div
  data-testid="accordion-toggle-icon"
  class="a b c d e f g h i j k l m n o accordion-toggle-icon"
></div>
;
"
`;

exports[`components/Accordion/ToggleIcon should render an open ToggleIcon 1`] = `
".a {
  flex-shrink: 0;
}
.b {
  width: 1.25rem;
}
.c {
  height: 1.25rem;
}
.d {
  margin-left: auto;
}
.e {
  background-image: url("/chevron-down.svg");
}
.f {
  background-repeat: no-repeat;
}
.g {
  background-size: 1.25rem;
}
.l {
  transform: rotate(-180deg);
}
.m {
  transition-property: transform;
}
.n {
  transition-duration: 0.2s;
}
.o {
  transition-delay: ease-in-out;
}
.p {
  transition-timing-function: ease;
}
@media (prefers-reduced-motion: reduce) {
  .h {
    transition-property: none;
  }
  .i {
    transition-duration: 0s;
  }
  .j {
    transition-delay: 0s;
  }
  .k {
    transition-timing-function: ease;
  }
}


<div
  data-testid="accordion-toggle-icon"
  class="a b c d e f g h i j k l m n o p accordion-toggle-icon"
></div>
;
"
`;
